E0185 needs to be updated to new format

From: src/test/compile-fail/E0185.rs

E0185 needs a span_label, updating it from:

```
error[E0185]: method `foo` has a `&self` declaration in the impl, but not in the trait
  --> src/test/compile-fail/E0185.rs:18:5
   |
18 |     fn foo(&self) {} //~ ERROR E0185
   |     ^^^^^^^^^^^^^^^^
```

To:

```
error[E0185]: method `foo` has a `&self` declaration in the impl, but not in the trait
  --> src/test/compile-fail/E0185.rs:18:5
   |
18 |     fn foo(&self) {} //~ ERROR E0185
   |     ^^^^^^^^^^^^^^^^ `&self` not used in the trait
```

Bonus: show the trait that it's referencing.

```
error[E0185]: method `foo` has a `&self` declaration in the impl, but not in the trait
  --> src/test/compile-fail/E0185.rs:18:5
   |
12 |     fn foo();
   |     --------- trait declared without `&self`
...
18 |     fn foo(&self) {} //~ ERROR E0185
   |     ^^^^^^^^^^^^^^^^ `&self` used in impl
```

--- a/src/librustc_typeck/check/compare_method.rs
+++ b/src/librustc_typeck/check/compare_method.rs
@@ -59,19 +59,33 @@ pub fn compare_impl_method<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
         (&ty::ExplicitSelfCategory::Static,
          &ty::ExplicitSelfCategory::Static) => {}
         (&ty::ExplicitSelfCategory::Static, _) => {
-            span_err!(tcx.sess, impl_m_span, E0185,
+            let mut err = struct_span_err!(tcx.sess, impl_m_span, E0185,
                 "method `{}` has a `{}` declaration in the impl, \
                         but not in the trait",
                         trait_m.name,
                         impl_m.explicit_self);
+            err.span_label(impl_m_span, &format!("`{}` used in impl",
+                                                 impl_m.explicit_self));
+            if let Some(span) = tcx.map.span_if_local(trait_m.def_id) {
+                err.span_label(span, &format!("trait declared without `{}`",
+                                              impl_m.explicit_self));
+            }
+            err.emit();
             return;
         }
         (_, &ty::ExplicitSelfCategory::Static) => {
-            span_err!(tcx.sess, impl_m_span, E0186,
+            let mut err = struct_span_err!(tcx.sess, impl_m_span, E0186,
                 "method `{}` has a `{}` declaration in the trait, \
                         but not in the impl",
                         trait_m.name,
                         trait_m.explicit_self);
+            err.span_label(impl_m_span, &format!("expected `{}` in impl",
+                                                  trait_m.explicit_self));
+            if let Some(span) = tcx.map.span_if_local(trait_m.def_id) {
+                err.span_label(span, & format!("`{}` used in trait",
+                                               trait_m.explicit_self));
+            }
+            err.emit();
             return;
         }
         _ => {

--- a/src/test/compile-fail/E0185.rs
+++ b/src/test/compile-fail/E0185.rs
@@ -9,13 +9,14 @@
 // except according to those terms.
 
 trait Foo {
-    fn foo();
+    fn foo(); //~ trait declared without `&self`
 }
 
 struct Bar;
 
 impl Foo for Bar {
     fn foo(&self) {} //~ ERROR E0185
+    //~^ `&self` used in impl
 }
 
 fn main() {

--- a/src/test/compile-fail/E0186.rs
+++ b/src/test/compile-fail/E0186.rs
@@ -9,13 +9,14 @@
 // except according to those terms.
 
 trait Foo {
-    fn foo(&self);
+    fn foo(&self); //~ `&self` used in trait
 }
 
 struct Bar;
 
 impl Foo for Bar {
     fn foo() {} //~ ERROR E0186
+    //~^ expected `&self` in impl
 }
 
 fn main() {

